dev-clean/headers/llsc/value_ops.hpp

```diff
@@ -93,23 +145,30 @@ struct std::equal_to<PtrVal> {
   }
 };
 
-template<>
-struct std::equal_to<immer::flex_vector<PtrVal>> {
-  bool operator()(immer::flex_vector<PtrVal> const& a, immer::flex_vector<PtrVal> const& b) const {
-    if (a.size() != b.size()) return false;
-    for (int i = 0; i < a.size(); i++)
-      if (a.at(i).get() != b.at(i).get()) return false;
-    return true;
-  }
-};
+// template<>
+// struct std::equal_to<immer::flex_vector<PtrVal>> {
+//   bool operator()(immer::flex_vector<PtrVal> const& a, immer::flex_vector<PtrVal> const& b) const {
+//     if (a.size() != b.size()) return false;
+//     for (int i = 0; i < a.size(); i++)
+//       if (a.at(i).get() != b.at(i).get()) return false;
+//     return true;
+//   }
+// };
 
 inline phmap::parallel_flat_hash_set<PtrVal,
-    std::hash<PtrVal>,
-    std::equal_to<PtrVal>,
+    hash_PtrVal,
+    equal_to_PtrVal,
     std::allocator<PtrVal>,
     4,
     std::mutex> objpool;
 
+inline PtrVal hashconsing(const PtrVal &ret) {
```

Comment:
We could do a lookup first and return the cached one if hit (thus avoid an allocation/deletion when missed)? 

And when do we need to delete the object memory if not using shared_ptr? It seems we need to implement some form of GC.

Reply:
In the `echo` case, `hashconsing` is taking about 4.5%, and other `make_*` calls are also taking modest time. There's no urgent need to avoid the allocation/deletion pair, as the current implementation has the advantage of being scalable wrt the kinds of values. Nevertheless, memory pooling is definitely an ideal solution.

Currently, when using `simple_ptr`s, we should delete an object if there's an identical object found in the global `objpool`. Ideally, we can use a memory pool, and all the memory allocation/freeing would happen at one end of the pool. The challenge would be to synchronize this pool (or multiple pools) across threads.

Reply:
For now, I'm not intending to include a memory pool implementation in this PR. Actually, I am not pretty sure how we can make thread-local storage efficient to access. We just fixed the `checker_map` performance issue [in a really strange way](https://github.com/Kraks/sai/pull/112/commits/d719b2acb901071a42e6f9f1a9ff657d0b1d4c1b), and I guess we should spend more time looking into `thread_local` in C++ before making further tweaks.

Reply:
From my perspective, the latter case is not really necessary, in the case where `IntV(1)` may be created again at a later point.

Reply:
If that makes sense to you, I'm intending to merge this PR now, and plan another PR for optimizing `IntV` and the memory pooling stuff.

Reply:
But if IntV(1) (and any other kinds of values) is not going to be used/recreated later, we have memory leaks. You can think that our backend implementation now gets closer to a VM, and this `objpool` is the heap of the VM, it cannot be only monotonically increasing in its content (if I didn't miss anything).

We can discuss more about memory management issue later and merge this PR first.

Reply:
I see your point. I was like letting it leak, but if we are constantly making new & different integers, we will be in trouble.

However, things aren't made worse by this PR, as we were not using a weak map for `objpool` in the beginning. I'm merging this now and will have a look at the integer issue.